src/common/service/NetRemoteDataStreamingReactors.cxx

```diff
@@ -107,19 +123,33 @@ DataStreamWriter::OnWriteDone(bool isOk)
 void
 DataStreamWriter::OnCancel()
 {
+    LOGD << "Enter OnCancel";
+
     // The RPC is cancelled by the client, so call Finish to complete it from the server perspective.
-    Finish(grpc::Status::CANCELLED);
+    if (!m_isCancelled.load(std::memory_order_relaxed)) {
+        m_isCancelled.store(true, std::memory_order_relaxed);
+        Finish(grpc::Status::CANCELLED);
+    }
 }
 
 void
 DataStreamWriter::OnDone()
 {
+    LOGD << "Enter OnDone";
     delete this;
 }
 
 void
 DataStreamWriter::NextWrite()
 {
+    LOGD << "Enter NextWrite";
+
+    // Client may have cancelled the RPC, so check for cancellation to prevent writing more data
+    // when we shouldn't.
+    if (m_isCancelled.load(std::memory_order_relaxed)) {
+        return;
```

Comment:
Maybe do a `LOGD` here indicating the write was aborted due to the RPC being cancled?